x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizerTests.java

```diff
@@ -3312,4 +3366,279 @@ public static RLike rlike(Expression left, String exp) {
+
+            new Equals(EMPTY, getFieldAttribute("a", genericType), getFieldAttribute("b", genericType)),
+            new Equals(EMPTY, new Literal(EMPTY, 1, INTEGER), new Literal(EMPTY, List.of(1, 2, 3), INTEGER))
+        );
+        for (Expression item : items) {
+            Expression isNull = new IsNull(EMPTY, item);
+            Expression transformed = rule.rule(isNull);
+            assertEquals(isNull, transformed);
+
+            IsNotNull isNotNull = new IsNotNull(EMPTY, item);
+            transformed = rule.rule(isNotNull);
+            assertEquals(isNotNull, transformed);
+        }
+
+    }
+
+    //
+    // Propagate nullability (IS NULL / IS NOT NULL)
+    //
+
+    // a IS NULL AND a IS NOT NULL => false
+    public void testIsNullAndNotNull() throws Exception {
+        FieldAttribute fa = getFieldAttribute("a");
+
+        And and = new And(EMPTY, new IsNull(EMPTY, fa), new IsNotNull(EMPTY, fa));
+        assertEquals(FALSE, new LogicalPlanOptimizer.PropagateNullable().rule(and));
+    }
+
+    // a IS NULL AND b IS NOT NULL AND c IS NULL AND d IS NOT NULL AND e IS NULL AND a IS NOT NULL => false
+    public void testIsNullAndNotNullMultiField() throws Exception {
+        FieldAttribute fa = getFieldAttribute("a");
+
+        And andOne = new And(EMPTY, new IsNull(EMPTY, fa), new IsNotNull(EMPTY, getFieldAttribute("a")));
+        And andTwo = new And(EMPTY, new IsNull(EMPTY, getFieldAttribute("a")), new IsNotNull(EMPTY, getFieldAttribute("a")));
+        And andThree = new And(EMPTY, new IsNull(EMPTY, getFieldAttribute("a")), new IsNotNull(EMPTY, fa));
+
+        And and = new And(EMPTY, andOne, new And(EMPTY, andThree, andTwo));
+
+        assertEquals(FALSE, new LogicalPlanOptimizer.PropagateNullable().rule(and));
+    }
+
+    // a IS NULL AND a > 1 => a IS NULL AND false
+    public void testIsNullAndComparison() throws Exception {
+        FieldAttribute fa = getFieldAttribute("a");
+        IsNull isNull = new IsNull(EMPTY, fa);
+
+        And and = new And(EMPTY, isNull, greaterThanOf(fa, ONE));
+        assertEquals(new And(EMPTY, isNull, nullOf(BOOLEAN)), new LogicalPlanOptimizer.PropagateNullable().rule(and));
+    }
+
+    // a IS NULL AND b < 1 AND c < 1 AND a < 1 => a IS NULL AND b < 1 AND c < 1 => a IS NULL AND b < 1 AND c < 1
+    public void testIsNullAndMultipleComparison() throws Exception {
+        FieldAttribute fa = getFieldAttribute("a");
+        IsNull isNull = new IsNull(EMPTY, fa);
+
+        And nestedAnd = new And(EMPTY, lessThanOf(getFieldAttribute("b"), ONE), lessThanOf(getFieldAttribute("c"), ONE));
+        And and = new And(EMPTY, isNull, nestedAnd);
+        And top = new And(EMPTY, and, lessThanOf(fa, ONE));
+
+        Expression optimized = new LogicalPlanOptimizer.PropagateNullable().rule(top);
+        Expression expected = new And(EMPTY, and, nullOf(BOOLEAN));
```

Comment:
Actually, running this through a test (`row a = 1, b = 0, c = 0 | eval a IS NULL AND b < 1 AND c < 1 AND a <= 1`) the plan is optimized as `[ISNULL(1[INTEGER]) AND null[BOOLEAN] AND null[BOOLEAN] AND null[BOOLEAN]`. This is not right. For whatever reason (I stopped looking into it at this point), `1` (as Literal) is added to the list of `nullExpressions` in `OptimizerRules.PropagateNullable.rule` and then, with your change in `LogicalPlanOptimizer.PropagateNullable.nullify` the `1` (Literal) is replaced as the "null" value in all other expressions where 1 is encountered.

For a different test where the coincidence of `1` disappears (`a = 1` and `a is null` and `b < 1`.... that is the same `1`) the rule yields something else. So the test is `row a = 5, b = 0, c = 0 | eval a IS NULL AND b < 1 AND c < 1 AND a <= 1` and the rule PropagateNullable changes the plan to `[ISNULL(5[INTEGER]) AND 0[INTEGER] < 1[INTEGER] AND 0[INTEGER] < 1[INTEGER] AND null[BOOLEAN]`.

I see two problems, that all might have to do with a single source:
- `1` gets replaced in the expressions as `null`
- `a <= 1` becomes `null`

Reply:
That's interesting, I'll look deeper into this case, maybe I broke something in the original logic

Reply:
The problem seems to be due to an interference between two rules ([PropagateEvalFoldables](https://github.com/elastic/elasticsearch/blob/main/x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizer.java#L479) and [PropagateNullable](https://github.com/elastic/elasticsearch/blob/main/x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/optimizer/OptimizerRules.java#L1705)), my change only exposed it. 
In addition, the different semantics of nullability in ESQL makes this whole management much more complex.

What happens is the following:
1. `PropagateEvalFoldables` replaces all the occurrences of `a` and `b` with `1`, so `row a = 1, b = 1, c = 1 | eval a IS NULL AND b < 1 AND c < 1 AND a <= 1` becomes `row a = 1, b = 1, c = 1 | eval 1 IS NULL AND 1 < 1 AND 1 < 1 AND 1 <= 1`. This is not evident in the planning logs, because when we print the single plans we only see the original source, not the actual values in the expression nodes (see screenshot below)
2. `PropagateNullable` tries to optimize `a IS NULL ... AND a <= 1` as `a IS NULL ... AND NULL <= 1` (and not `a IS NULL AND FALSE`, the comment [here](https://github.com/elastic/elasticsearch/blob/main/x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/optimizer/OptimizerRules.java#L1703) is wrong. The replacer used in the rule is [this::nullify](https://github.com/elastic/elasticsearch/blob/main/x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/optimizer/OptimizerRules.java#L1741), that produces `NULL` for the attribute, not `FALSE` for the whole expression). 
3. This happens by [finding the field](https://github.com/elastic/elasticsearch/blob/main/x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/optimizer/OptimizerRules.java#L1722) that is supposed to be null (`a` in this case), looking for it in other expressions, and [replacing it](https://github.com/elastic/elasticsearch/blob/main/x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/optimizer/OptimizerRules.java#L1741) with `NULL`. The problem is that there is no `a` at this time, because it was replaced with `1` (see point 1). Even worse, also other expressions were replaced with `1` (in this case, `b < 1` is now `NULL < 1`), so they will also be affected by this replacement, that is completely unexpected.

See this screenshot: when trying to infer the null expressions, the actual field value `isn.field()`, that was expected to return `a`, returns `1` instead

<img width="683" alt="Screenshot 2024-03-05 alle 14 08 16" src="https://github.com/elastic/elasticsearch/assets/459289/69cca2b0-080d-4d05-aa48-90b168db58a2">

---

So why did it work before?

In the original implementation it worked only because expressions like `constant < constant` were considered non-nullable, so `nullify()` returned the original expression. In practice, the rule did nothing in this case.

IMHO this is another problem: [relying on nullability to replace values with NULL](https://github.com/elastic/elasticsearch/blob/main/x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/optimizer/OptimizerRules.java#L1782) does not seem to be so safe: if we know that `a == 1` and we try to optimize `a < 0`, we cannot just replace `a` with null, because calculating  `1 < 0` is different from calculating `null < 0`, due to null virality.

--- 

How to fix it?

The problem seems to be limited to foldable literals, so we could explicitly exclude this case. Something like

```java
        protected Expression nullify(Expression exp, Expression nullExp) {
            return exp.foldable() ? exp : Literal.of(exp, null);
        }
```
An alternative is to completely drop this rule, that would be the safest path, but also a shame because it looks correct (though a bit intricate)  for non-foldable attributes, so we would give up on a valid optimization.

Or we could try to completely rewrite it, but it doesn't seem trivial.

Suggestions are welcome

Reply:
My alternative from a month ago was to split FoldNull: https://github.com/elastic/elasticsearch/compare/main...astefan:elasticsearch:fold_null_split?expand=1.

Reply:
Turns out that the fix above on `nullify()` method is not enough: if a constant value becomes a target for "nullification", it will be transformed in other unexpected places, eg. in the right-hand side of a completely unrelated expression.

I went for a different solution: exclude constant expressions from `PropagateNullable` logic since the beginning. 
In terms of performance it should not make any difference, foldable expressions will eventually be folded anyway, there is no need to nullify them in advance.
The case you mentioned above should be fine now, I added a specific test for that

Reply:
> The replacer used in the rule is [this::nullify](https://github.com/elastic/elasticsearch/blob/main/x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/optimizer/OptimizerRules.java?rgh-link-date=2024-03-05T15%3A05%3A13Z#L1741), that produces NULL for the attribute, not FALSE for the whole expression).

The original method returns null IF the expression is nullable which is true for functions only if at least one child is null. However it is also true for attributes / FieldAttributes leading to its replacement with null.
In might be that in the SQL planner this didn't get a chance kick in because of the other rules - one fix would be to change the nullify method.